File: ambari-server/src/main/java/org/apache/ambari/server/state/Service.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.ambari.server.state;

import java.util.List;
import java.util.Map;
import java.util.Set;

import org.apache.ambari.annotations.Experimental;
import org.apache.ambari.annotations.ExperimentalFeature;
import org.apache.ambari.server.AmbariException;
import org.apache.ambari.server.api.services.ServiceKey;
import org.apache.ambari.server.controller.ServiceDependencyResponse;
import org.apache.ambari.server.controller.ServiceResponse;
import org.apache.ambari.server.orm.entities.ClusterServiceEntity;
import org.apache.ambari.server.orm.entities.RepositoryVersionEntity;

public interface Service {

  Long getClusterId();

  Long getServiceGroupId();

  String getServiceGroupName();

  Long getServiceId();

  String getName();

  String getDisplayName();

  String getServiceType();

  Cluster getCluster();

  Set<ServiceDependencyResponse> getServiceDependencyResponses();

  List<ServiceKey> getServiceDependencies();

  ServiceComponent getServiceComponent(String componentName)
      throws AmbariException;

  Map<String, ServiceComponent> getServiceComponents();

  void addServiceComponents(Map<String, ServiceComponent> components)
      throws AmbariException;

  void addServiceComponent(ServiceComponent component)
      throws AmbariException;

  State getDesiredState();

  void setDesiredState(State state);

  StackId getDesiredStackId();

  ServiceResponse convertToResponse();

  void debugDump(StringBuilder sb);

  ServiceComponent addServiceComponent(String serviceComponentName, String serviceComponentType)
      throws AmbariException;

  /**
   * Find out whether the service and its components
   * are in a state that it can be removed from a cluster
   * @return
   */
  boolean canBeRemoved();

  void deleteAllComponents() throws AmbariException;

  void deleteServiceComponent(String componentName)
      throws AmbariException;

  boolean isClientOnlyService();

  ClusterServiceEntity removeDependencyService(Long dependencyServiceId);

  ClusterServiceEntity addDependencyService(Long dependencyServiceId) throws AmbariException;

  void delete() throws AmbariException;

  /**
   * Sets the maintenance state for the service
   * @param state the state
   */
  void setMaintenanceState(MaintenanceState state);

  /**
   * @return the maintenance state
   */
  MaintenanceState getMaintenanceState();

  /**
   * Refresh Service info due to current stack
   * @throws AmbariException
   */
  void updateServiceInfo() throws AmbariException;


  /**
   * Get a true or false value specifying
   * whether credential store is supported by this service.
   * @return true or false
   */
  boolean isCredentialStoreSupported();

  /**
   * Get a true or false value specifying
   * whether credential store is required by this service.
   * @return true or false
   */
  boolean isCredentialStoreRequired();

  /**
   * Get a true or false value specifying whether
   * credential store use is enabled for this service.
   *
   * @return true or false
   */
  boolean isCredentialStoreEnabled();

  /**
   * Set a true or false value specifying whether this
   * service is to be enabled for credential store use.
   *
   * @param credentialStoreEnabled - true or false
   */
  void setCredentialStoreEnabled(boolean credentialStoreEnabled);

  /**
   * @return
   */
  @Deprecated
  @Experimental(feature = ExperimentalFeature.REPO_VERSION_REMOVAL)
  RepositoryVersionEntity getDesiredRepositoryVersion();

  /**
   * @param desiredRepositoryVersion
   */
  @Deprecated
  @Experimental(feature = ExperimentalFeature.REPO_VERSION_REMOVAL)
  void setDesiredRepositoryVersion(RepositoryVersionEntity desiredRepositoryVersion);

  /**
   * Gets the repository for the desired version of this service by consulting
   * the repository states of all known components.
   */
  @Deprecated
  @Experimental(feature = ExperimentalFeature.REPO_VERSION_REMOVAL)
  RepositoryVersionState getRepositoryState();

  enum Type {
    HDFS,
    GLUSTERFS,
    MAPREDUCE,
    HBASE,
    HIVE,
    OOZIE,
    WEBHCAT,
    SQOOP,
    GANGLIA,
    ZOOKEEPER,
    PIG,
    FLUME,
    YARN,
    MAPREDUCE2,
    AMBARI_METRICS,
    KERBEROS
  }
}

File: ambari-server/src/main/resources/stacks/HDP/2.1.GlusterFS/services/YARN/package/scripts/service.py
"""
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Ambari Agent

"""

from resource_management import *


def service(componentName, action='start', serviceName='yarn'):

  import params

  if (serviceName == 'mapreduce' and componentName == 'historyserver'):
    daemon = format("{mapred_bin}/mr-jobhistory-daemon.sh")
    pid_file = format("{mapred_pid_dir}/mapred-{mapred_user}-{componentName}.pid")
    usr = params.mapred_user
  else:
    daemon = format("{yarn_bin}/yarn-daemon.sh")
    pid_file = format("{yarn_pid_dir}/yarn-{yarn_user}-{componentName}.pid")
    usr = params.yarn_user

  cmd = format("export HADOOP_LIBEXEC_DIR={hadoop_libexec_dir} && {daemon} --config {config_dir}")

  if action == 'start':
    daemon_cmd = format("{cmd} start {componentName}")
    no_op = format("ls {pid_file} >/dev/null 2>&1 && ps -p `cat {pid_file}` >/dev/null 2>&1")
    Execute(daemon_cmd,
            user=usr,
            not_if=no_op
    )

    Execute(no_op,
            user=usr,
            not_if=no_op,
            initial_wait=5
    )

  elif action == 'stop':
    daemon_cmd = format("{cmd} stop {componentName}")
    Execute(daemon_cmd,
            user=usr,
    )
    rm_pid = format("rm -f {pid_file}")
    Execute(rm_pid,
            user=usr
    )
